### ISSUE TITLE
Send tracking info back to Reverb

### ISSUE THREAD

**Issue by skwp** [reverbdotcom/reverb-magento#68]
When a shipment is created in magento / tracking info entered, ping the Reverb API with the tracking info for this shipment. The link for this is available from order._links.ship.href in the original order details and looks like "my/orders/selling/#{id}/ship". It's better to read the link rather than hardcode it 

When you POST to this url /my/orders/selling/id/ship, the followoing params are required:

`provider`: valid : UPS, USPS, FedEx, DHL, Canada Post, Other
If we can't figure out from magento which one it is, use "Other"

`tracking_number` - the tracking number for the carrier
`send_notification` -  set to `true`. this is so we notify the buyer

Queue Task objects for the shipment-tracking communication should be accessible in the same admin grid as order creation. This page/grid should be renamed to indicate such

The following combination of values will only be sent to Reverb ONCE (unless admins manually interact with the queue object to resend). No reason the same set of values should need to be re-communicated
reverb_order_id
shipping carrier code
tracking_number

- Correct that we don't have to send numbers more than once
- If a number is deleted there is no corresponding action to take on Reverb as we would have already marked it as shipped
 - If a number is updated we should send it to Reverb

Valid carriers
"UPS"
"USPS"
"FedEx"
"DHL"
"Canada Post"
"Other"

**Comment by dunagan5887** [comment]
Where does this rank in terms of priority?

**Comment by skwp** [comment]
after all the minor order sync cleanup things but before image and category sync 

--
Yan Pritzker
yanpritzker.com
via iPhone

**Comment by dunagan5887** [comment]
So the Magento admin provides the Shipping Provider and the Tracking Number, correct? And upon creation of the shipment_tracking object in the Magento system, this post is made as an event observer, correct? If the Reverb API call fails (due to Reverb API server being temporarily down or other issues) should Magento still allow creation of the shipment_tracking object?

**Comment by dunagan5887** [comment]
I would estimate this at 4-8 hours

**Comment by skwp** [comment]
sounds good. the reverb api call should be retried in the background if possible. we don't want it to hold up anything on the magento side

**Comment by skwp** [comment]
if that adds significant complexity we can try to make it inline with the save of the magento shipment as the reverb API is generally up with good uptime

**Comment by dunagan5887** [comment]
Sounds like it would be better to add a queue item as part of the Magento-tracking_shipment-object's save transaction. This would ensure that the task is in the system but allow it to be requeued. 

Is there any reason that a tracking number would need to be sent multiple times to Reverb?

**Comment by skwp** [comment]
Should be only once as long as it succeeds (api is up, etc)

**Comment by dunagan5887** [comment]
Ok I'll make the unique key value be a concatenation of the magento_order_id, magento_shipment_id, carrier_code and tracking number. This will ensure that the tracking number is only sent once, even if the admin accidentally adds a second shipment_tracking row with the same tracking number. In this event, I won't prevent the Magento shipment_tracking number from being completed

**Comment by dunagan5887** [comment]
Also if we're going to need a grid to allow shipment tracking queue items to be retried, that would take another 2 hours or so

**Comment by skwp** [comment]
actually if someone changes the tracking number for example because they did a typo the first time we _do_ want to send that to reverb.

**Comment by skwp** [comment]
UPS = "UPS"
  USPS = "USPS"
  FEDEX = "FedEx"
  DHL = "DHL"
  CANADA_POST = "Canada Post"
  OTHER = "Other"

**Comment by skwp** [comment]
Only the strings are important, i'l update the above to reduce confusion

**Comment by dunagan5887** [comment]
Estimating this at 8-12 hours to get the base functionality working; UX refactors will add time to this estimate.

@skwp  "actually if someone changes the tracking number for example because they did a typo the first time we do want to send that to reverb."
So let's say the admin deletes the initial typo tracking number, should an API call be made to Reverb to indicate that the initial tracking number was deleted? 

Also, in terms of the grid these will show up in, should they have their own distinct grid? Or should they combined with the Order Creation table? Also its important now that we establish the following technical requirement:

The following combination of values will only be sent to Reverb ONCE (unless admins manually interact with the queue object to resend):
reverb_order_id
shipping carrier code
tracking_number
In other words, there should be no reason to communicate a tracking number twice, correct?

**Comment by skwp** [comment]
1. Correct that we don't have to send numbers more than once
2. If a number is deleted there is no  corresponding action to take on Reverb as we would have already marked it as shipped
3. If a number is updated we should send it to Reverb

I think it would be ok to show this in the order update grid unless you think that's confusing. I'll go with your decision on that one.

**Comment by dunagan5887** [comment]
Due to the requirement of "Correct that we don't have to send numbers more than once" the queue task objects for this API process will reside in the same table as Order Creation. Having them in the order creation grid would be easy, having them in the order update grid would not be worth the time to implement in my opinion. The realistic choices here are either in a separate grid or with the Order Creation grid; we could refactor the name of that grid to be Order/Shipment Creation Grid

**Comment by skwp** [comment]
ok sounds good on the consolidation and rename

**Comment by dunagan5887** [comment]
@skwp  Also it is possible to create a shipment in Magento without entering tracking information. A queue task object should only be created if tracking information is entered, correct?

**Comment by dunagan5887** [comment]
Also, in the event that the attempt to create a queue task object for the tracking information fails (e.g. a develop has locked the queue table accidentally) should the tracking information still be allowed to be created on the shipment? Again, there's no reason that the queue task would fail to be created normally, this is just in extreme edge cases where there's interference with the instance

**Comment by skwp** [comment]
yes we only want to send tracking info if it's entered. that seems like extreme edge case behavior. however I don't think we should interfere with normal magento flows otherwise our extension impedes their usage. if we can't write our task then we'll just lose tracking info for that order and  they have to enter it manually later

**Comment by dunagan5887** [comment]
@skwp Yan, I'm currently getting a 404 when attempting to send a shipment tracking API POST to Reverb:

https://sandbox.reverb.com/my/orders/selling/117/ship
Status 404

I'll attach a screenshot of my debugger as well as the result from the API cURL request (404.jpg)

**Comment by dunagan5887** [comment]
![debugger](https://cloud.githubusercontent.com/assets/1616312/10039274/e0e28c36-6195-11e5-9c2a-408cfc2e7c72.png)

<img width="1468" alt="404_response" src="https://cloud.githubusercontent.com/assets/1616312/10039278/ea987362-6195-11e5-9bb2-387c95b4fd9e.png">

**Comment by dunagan5887** [comment]
Branch issues/github_68 is now ready to test the Shipment Tracking API call.

Looking at the legacy code, it's not overly clear how error responses from the server are supposed to be parsed. If error messaging is returned as part of the response, what form will the response take?

**Comment by skwp** [comment]
@dunagan5887 errors will look like this: http://dev.reverb.com/docs/error-handling

**Comment by skwp** [comment]
this particular API actually always succeeds right now. if it does return errors it will be in the format as shown above

**Comment by dunagan5887** [comment]
Sounds good, I'll code the response handling according to the link above. We will probably want to refactor the response handling for the legacy calls given that they are hard-coding response codes (which the link above expressly advises against).

In terms of the 404 being returned, is the specified route defined as a POST on the Reverb side?

**Comment by skwp** [comment]
yes it's a POST to /my/orders/selling/:id/ship

**Comment by skwp** [comment]
Sorry I'm just catching up on your messages, looking into it

**Comment by dunagan5887** [comment]
Any preference as to what I should do with the response (looks to be the same response as an order update call)? Currently I'm just going to update the status message on the queue task row to be "Shipment Tracking has been synced with Reverb"

**Comment by dunagan5887** [comment]
![shipment_tracking_response](https://cloud.githubusercontent.com/assets/1616312/10060771/113d8d08-6219-11e5-8b5d-bc42989eeb0c.png)
